File: examples/Haskell/peering1.hs
{-# LANGUAGE OverloadedLists   #-}
{-# LANGUAGE OverloadedStrings #-}

module Main where

import           Control.Monad         (forever, when)

import qualified Data.ByteString.Char8 as C
import           Data.Semigroup        ((<>))

import           System.Environment
import           System.Exit
import           System.Random

import           System.ZMQ4.Monadic

connectPeer :: Socket z t -> String -> String -> ZMQ z ()
connectPeer sock name peer = do
    connect sock (connectString peer name)
    liftIO . putStrLn $ "Connecting to peer: " ++ connectString peer name

connectString :: String -> String -> String
connectString peer name = "ipc://" ++ peer ++ "-" ++ name ++ ".ipc"

main :: IO ()
main = do
    args <- getArgs
    when (length args < 2) $ do
        putStrLn "Usage: peering1 <me> <you> [<you> ...]"
        exitFailure

    let self:peers = args
    putStrLn $ "Preparing broker at " ++ self

    runZMQ $ do
        -- Bind state backend to endpoint
        stateBack <- socket Pub
        bind stateBack (connectString self "state")

        -- Connect state frontend to peers
        stateFront <- socket Sub
        subscribe stateFront ""
        mapM_ (connectPeer stateFront "state") peers

        -- Send status, collect status
        forever $ do
            let pollItem = Sock stateFront [In] (Just pollEvent)
                pollEvent _ = do
                    peerName:available:_ <- receiveMulti stateFront
                    liftIO . C.putStrLn $
                        peerName <> " " <> available <> " workers free"
            pollEvents <- poll oneSec [pollItem]
            when (pollEvents == [[]]) $ do
                r <- liftIO $ randomRIO (0, 9)
                sendMulti stateBack [C.pack self, C.pack (show (r :: Int))]
  where
    oneSec = 1000

File: examples/Node.js/peering1.js
//  Broker peering simulation (part 1)
//  Prototypes the state flow

var zmq = require('zeromq')
  , util = require('util');

if (process.argv.length < 3) {
	console.log('usage: node peering1.js me [you ...]');
	process.exit(0);
}

var self = process.argv[2];
console.log("I: preparing broker at %s…", self);

// flag for stopping timer
var done = false;

//
// Backend
//

var statebe = zmq.socket('pub');
statebe.bindSync(util.format("ipc://%s-state.ipc", self));

//
// Frontend
//

var statefe = zmq.socket('sub');
statefe.subscribe('');
for (var i = 3; i < process.argv.length; i++) {
	var peer = process.argv[i];
	console.log("I: connecting to state backend at '%s'", peer);
	statefe.connect(util.format("ipc://%s-state.ipc", peer));
}

process.on('SIGINT', function() {
  done = true;
  statebe.close();
  statefe.close();
});

//  The main loop sends out status messages to peers, and collects
//  status messages back from peers. 

statefe.on('message', function(peer_name, available) {
	console.log("%s - %s workers free", peer_name, available);
});

function sendWorkerAvailability() {
	if (done) {
		return;
	}
	var num_workers = util.format("%d", Math.floor(10 * Math.random()));
	console.log("sending update: %s has %s", self, num_workers);
	statebe.send([ self, num_workers ]);

	var next_send_delay = Math.floor(3000 * Math.random());
	setTimeout(sendWorkerAvailability, next_send_delay);
}

// Start worker update timer loop
sendWorkerAvailability();